### PR TITLE
Parse errors into severity

### PR DESCRIPTION
**Pull request by avilevy18**
## Description

Parse severity from Fluent Bit self logs

## Related issue

b/265979940

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [x] Integration tests do not apply.
  - [ ] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

File: confgenerator/confgenerator.go [hunk 1]
```diff
@@ -32,6 +32,8 @@ import (
 	"github.com/shirou/gopsutil/host"
 )
 
+const FluentBitSelfLogTag = "ops-agent-fluent-bit"
```
Comment (jefferbrecht) [inline confgenerator/confgenerator.go:35]: This should be unexported

File: confgenerator/logging_processors.go [hunk 2]
```diff
@@ -135,6 +135,7 @@ type LoggingProcessorParseRegex struct {
 	ConfigComponent `yaml:",inline"`
 	ParserShared    `yaml:",inline"`
 	Field           string `yaml:"field,omitempty"`
+	PreserveKey     bool   `yaml:"preserve_key,omitempty"`
```
Comment (jefferbrecht) [inline confgenerator/logging_processors.go:138]: Was it intentional to include this as an exposed config in the YAML? If so, are we going to leave it undocumented for now?

Comment (avilevy18) [reply confgenerator/logging_processors.go]: Changed in: [8145132](https://github.com/GoogleCloudPlatform/ops-agent/pull/1040/commits/81451323dc4ba31fa2296dcc8c76b75838800967)

File: confgenerator/confgenerator.go (hunk 1 above)
Comment (avilevy18) [reply confgenerator/confgenerator.go]: Changed in: [8145132](https://github.com/GoogleCloudPlatform/ops-agent/pull/1040/commits/81451323dc4ba31fa2296dcc8c76b75838800967)

File: confgenerator/confgenerator.go [hunk 3]
```diff
@@ -395,12 +396,16 @@ func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.I
 
 func generateSeveritySelfLogsParser() []fluentbit.Component {
 	out := make([]fluentbit.Component, 0)
+
+	_, timeOffset := time.Now().Zone()
```
Comment (jefferbrecht) [inline confgenerator/confgenerator.go:400]: `timeOffset` is in seconds here, but `Time_Offset` in fluent-bit is expecting a `+-hhmm`-style offset. Leading zero is also required. Something like this should work:

```go
timeOffset := time.Now().Format("-0700")
```

File: confgenerator/confgenerator.go [hunk 4]
```diff
@@ -395,12 +396,16 @@ func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.I
...
+
 	parser := LoggingProcessorParseRegex{
 		Regex:       `(?<message>\[[ ]*(?<time>\d+\/\d+\/\d+ \d+:\d+:\d+)] \[[ ]*(?<severity>[a-z]+)\].*)`,
 		PreserveKey: true,
 		ParserShared: ParserShared{
 			TimeKey:    "time",
 			TimeFormat: "%Y/%m/%d %H:%M:%S",
+			TimeOffset: fmt.Sprintf("%d", timeOffset),
```
Comment (jefferbrecht) [inline confgenerator/confgenerator.go:408]: In addition to the above, I also think we should omit this if it's zero to simplify the diff.

File: confgenerator/fluentbit/processors.go [hunk 5]
```diff
@@ -110,6 +111,9 @@ func ParserComponentBase(TimeFormat string, TimeKey string, Types map[string]str
 	if TimeKey != "" {
 		parser.Config["Time_Key"] = TimeKey
 	}
+	if TimeOffset != "" {
```
Comment (jefferbrecht) [inline confgenerator/fluentbit/processors.go:114]: Actually, the omit-if-zero idea might be more appropriate here instead:

```suggestion
	if TimeOffset != "" && TimeOffset != "+0000" && TimeOffset != "-0000" {
```

File: confgenerator/confgenerator.go (hunk 3 above)
Comment (avilevy18) [reply confgenerator/confgenerator.go:400]: Great catch, thanks!

File: confgenerator/confgenerator.go [hunk 6]
```diff
@@ -383,12 +386,55 @@ func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.I
...
+	out = append(out, fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Key_Name": "message",
+			"Name":     "parser",
+			"Parser":   "logging-severity-parser.self-logs-severity",
+			"Match":    fluentBitSelfLogTag,
+		},
+	})
+
```
Comment (LujieDuan) [inline confgenerator/confgenerator.go:426]: Question: Do we need the extra `FILTER` here? Can we let `LoggingProcessorParseRegex` match on `fluentBitSelfLogTag` directly? 
```suggestion
	}.Components(fluentBitSelfLogTag, "self-logs-severity")

	out = append(out, parser...)

```

**Review by LujieDuan — COMMENTED** [review]
Suggestion: we could add an integration test case to verify the severity got extracted correctly. I believe Fluent Bit always has those `info` log lines on start up, so maybe we could just ```gce.WaitForLog(..., "ops-agent-fluent-bit", ..., `severity="INFO"`)```.

File: integration_test/ops_agent_test.go [hunk 7]
```diff
@@ -2672,6 +2672,27 @@ func TestLoggingAgentCrashRestart(t *testing.T) {
...
+		ctx, logger, vm := agents.CommonSetup(t, platform)
+
+		if err := setupOpsAgent(ctx, logger, vm, ""); err != nil {
+			t.Fatal(err)
+		}
+
+		// Wait long enough for the data to percolate through the backends
+		// under normal circumstances. Based on some experiments, 2 minutes
+		// is normal; wait a bit longer to be on the safe side.
+		time.Sleep(3 * time.Minute)
```
Comment (jefferbrecht) [inline integration_test/ops_agent_test.go:2688]: Is this actually necessary in practice? `gce.WaitForLog` already waits for 400 seconds...

File: confgenerator/confgenerator.go (hunk 6 above)
Comment (avilevy18) [reply confgenerator/confgenerator.go]: Removed extra filter, and adding integration tests

File: integration_test/ops_agent_test.go (hunk 7 above)
Comment (avilevy18) [reply integration_test/ops_agent_test.go]: Good catch! Will remove

File: integration_test/ops_agent_test.go [hunk 8]
```diff
@@ -2688,6 +2688,30 @@ func TestLoggingAgentCrashRestart(t *testing.T) {
...
+
+		if err := setupOpsAgent(ctx, logger, vm, ""); err != nil {
+			t.Fatal(err)
+		}
+
+		cmd := `sudo timedatectl set-timezone America/Toronto`
+		if gce.IsWindows(platform) {
+			cmd = `Set-TimeZone -Name "Eastern Standard Time"`
+		}
+		if _, err := gce.RunRemotely(ctx, logger.ToMainLog(), vm, "", cmd); err != nil {
```
Comment (jefferbrecht) [inline integration_test/ops_agent_test.go:2705]: This is happening too late into the test: this changes the time zone at runtime which isn't what we want, and fluent-bit has already finished flushing its startup self-logs by this point anyway so you wouldn't be asserting on logs with the adjusted timestamp. The `setupOpsAgent` command should be run after changing the time zone.

**Review by jefferbrecht — APPROVED** [review]

**Review by LujieDuan — APPROVED** [review]
LGTM!

**Comment by quentinmit** [comment]
Won't these tests fail if the developer's workstation is not in EST? (Also... does it handle DST correctly?)

**Comment by jefferbrecht** [comment]
I've been using `TZ=UTC+0 go test ./...` to get around it, maybe we can just set that in the unit tests?

**Comment by quentinmit** [comment]
> I've been using `TZ=UTC+0 go test ./...` to get around it, maybe we can just set that in the unit tests?

I don't think that's reasonable to tell anyone, and besides, won't this still break in production during DST due to the hardcoded offset?

**Comment by jefferbrecht** [comment]
> > I've been using `TZ=UTC+0 go test ./...` to get around it, maybe we can just set that in the unit tests?
> 
> I don't think that's reasonable to tell anyone, and besides, won't this still break in production during DST due to the hardcoded offset?

I meant that we can just set the TZ variable in TestMain or something so that the tests run under UTC; the user doesn't have to do anything different.

I'm assuming that the DST offset is rolled into the offset that we emit to the config so I think it'd be fine, but it should still be checked.

**Comment by quentinmit** [comment]
> I'm assuming that the DST offset is rolled into the offset that we emit to the config so I think it'd be fine, but it should still be checked.

Nothing is going to restart fluent-bit every time the DST offset changes, so it will be wrong until the next time fluent-bit happens to restart.

**Pull request merged**